From: JiaY-shi <shi05275@163.com>
Date: Fri, 24 Jul 2026 16:30:00 +0800
Subject: [PATCH] wifi: ath12k: allow DT BDF address offset

Recent IPQ5332 firmware uses a larger BDF offset to accommodate platform
variants with additional userPD images. Older firmware uses the original
memory layout and expects BDF memory at the original offset.

Allow boards using such firmware to override the hardware default through
Device Tree. Keep the hardware parameter as the default for all other
devices.

Signed-off-by: JiaY-shi <shi05275@163.com>
---
 drivers/net/wireless/ath/ath12k/qmi.c | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -2737,6 +2737,7 @@ static int ath12k_qmi_assign_target_mem_
 {
 	struct reserved_mem *rmem;
 	size_t avail_rmem_size;
+	u32 bdf_addr_offset;
 	int i, idx, ret;
 
 	for (i = 0, idx = 0; i < ab->qmi.mem_seg_count; i++) {
@@ -2778,7 +2779,20 @@ static int ath12k_qmi_assign_target_mem_
 				goto out;
 			}
 
-			avail_rmem_size = rmem->size - ab->hw_params->bdf_addr_offset;
+			if (of_property_read_u32(ab->dev->of_node,
+						 "qcom,bdf-address-offset",
+						 &bdf_addr_offset))
+				bdf_addr_offset = ab->hw_params->bdf_addr_offset;
+
+			if (bdf_addr_offset >= rmem->size) {
+				ath12k_err(ab,
+					   "BDF offset 0x%x exceeds reserved size %pa\n",
+					   bdf_addr_offset, &rmem->size);
+				ret = -EINVAL;
+				goto out;
+			}
+
+			avail_rmem_size = rmem->size - bdf_addr_offset;
 			if (avail_rmem_size < ab->qmi.target_mem[i].size) {
 				ath12k_dbg(ab, ATH12K_DBG_QMI,
 					   "failed to assign mem type %u req size %u avail size %zu\n",
@@ -2789,7 +2803,10 @@ static int ath12k_qmi_assign_target_mem_
 				goto out;
 			}
 			ab->qmi.target_mem[idx].paddr =
-				rmem->base + ab->hw_params->bdf_addr_offset;
+				rmem->base + bdf_addr_offset;
+			ath12k_info(ab, "using BDF offset 0x%x addr %pa\n",
+				    bdf_addr_offset,
+				    &ab->qmi.target_mem[idx].paddr);
 			ab->qmi.target_mem[idx].v.ioaddr =
 				ioremap(ab->qmi.target_mem[idx].paddr,
 					ab->qmi.target_mem[i].size);
